src/main/java/org/rascalmpl/maven/AbstractRascalMojo.java

```diff
@@ -0,0 +1,380 @@
+			throw new RuntimeException(e);
+		}
+		catch (MojoExecutionException e) {
+			log.error("Unable to just-in-time-install the required (bootstrap) version of Rascal. Have to abort.");
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected List<Path> collectDependentArtifactLibraries(MavenProject project) throws URISyntaxException, IOException {
+		List<Path> libs = new LinkedList<>();
+
+		for (Object o : project.getArtifacts()) {
+			Artifact a = (Artifact) o;
+			File file = a.getFile().getAbsoluteFile();
+
+			libs.add(file.toPath());
+		}
+
+		return libs;
+	}
+
+	protected final ArtifactVersion getReferenceRascalVersion() {
+		return new DefaultArtifactVersion("0.41.0-RC16");
+	}
+
+	protected Path installBootstrapRascalVersion() throws MojoExecutionException {
+		// download the boostrap version from our maven site
+		executeMojo(
+			plugin(
+				"org.apache.maven.plugins",
+				"maven-dependency-plugin",
+				"3.1.1"
+			),
+			goal("get"),
+			configuration(
+      			  element("artifact", "org.rascalmpl:rascal:" + bootstrapRascalVersion)
+    		),
+			executionEnvironment(
+				project,
+				session,
+				pluginManager
+			)
+		);
+
+		// resolve the jar file in the .m2 repository
+		return Path.of(session.getSettings().getLocalRepository(),
+			"org", "rascalmpl", "rascal", bootstrapRascalVersion, "rascal-" + bootstrapRascalVersion + ".jar");
+	}
+
+	protected Process runMain(boolean verbose, List<Path> todoList, List<Path> srcs, List<Path> libs, Path generated, Path bin) throws IOException {
+		String javaHome = System.getProperty("java.home");
+        String javaBin = javaHome + File.separator + "bin" + File.separator + "java";
+
+		List<String> command = new LinkedList<String>();
+		command.add(javaBin);
+
+		System.getProperties().forEach((key, value) -> {
+			command.add("-D" + key + "=" + value);
+		});
+
+		// we put the entire pathConfig on the commandline, and finally the todoList for compilation.
+		command.add("-cp");
+		command.add(getRascalRuntime().toString());
+		command.add(mainClass);
+		if (mainModule != null && !mainModule.isEmpty()) {
+			command.add(mainModule);
+		}
+		command.add("-srcs");
+		command.add(srcs.stream().map(Object::toString).collect(Collectors.joining(File.pathSeparator)));
+		command.add("-libs");
+		command.add(libs.stream().map(Object::toString).collect(Collectors.joining(File.pathSeparator)));
+		command.add("-bin");
+		command.add(bin.toString());
+		command.add("-generatedSources");
+		command.add(generated.toString());
+		command.add("-modules");
+		command.add(todoList.stream().map(Object::toString).collect(Collectors.joining(File.pathSeparator)));
+		for (Entry<String, String> e : extraParameters.entrySet()) {
+			command.add("-" + e.getKey());
+			command.add(e.getValue());
+		}
+
+		if (verbose) {
+			command.add("-verbose");
+		}
+
+		return new ProcessBuilder(command).inheritIO().start();
```

Comment:
The plexus-tools from maven has some dedicated tooling to invoke nexted commands: https://codehaus-plexus.github.io/plexus-utils/apidocs/org/codehaus/plexus/util/cli/Commandline.html

can we use that? (and let it take care of some of the messy work?)

Reply:
noted. What messy work are you referring to? This is pretty straightforward stuff.